lxd/device/proxy.go

```diff
@@ -0,0 +1,410 @@
+package device
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"golang.org/x/sys/unix"
+	"gopkg.in/lxc/go-lxc.v2"
+
+	"github.com/lxc/lxd/lxd/device/config"
+	"github.com/lxc/lxd/lxd/instance"
+	"github.com/lxc/lxd/lxd/iptables"
+	"github.com/lxc/lxd/lxd/project"
+	"github.com/lxc/lxd/shared"
+)
+
+type proxy struct {
+	deviceCommon
+}
+
+type proxyProcInfo struct {
+	listenPid      string
+	connectPid     string
+	connectAddr    string
+	listenAddr     string
+	listenAddrGID  string
+	listenAddrUID  string
+	listenAddrMode string
+	securityUID    string
+	securityGID    string
+	proxyProtocol  string
+}
+
+func (d *proxy) CanHotPlug() (bool, []string) {
+	return true, []string{}
+}
+
+// validateConfig checks the supplied config for correctness.
+func (d *proxy) validateConfig() error {
+	if d.instance.Type() != instance.TypeContainer {
+		return ErrUnsupportedDevType
+	}
+
+	rules := map[string]func(string) error{
+		"listen":         shared.IsNotEmpty,
+		"connect":        shared.IsNotEmpty,
+		"bind":           shared.IsAny,
+		"mode":           shared.IsAny,
+		"gid":            shared.IsAny,
+		"uid":            shared.IsAny,
+		"nat":            shared.IsAny,
+		"security.uid":   shared.IsAny,
+		"security.gid":   shared.IsAny,
+		"proxy_protocol": shared.IsAny,
```

Comment:
We should do a better job at validating things here.

 - Validate that listen and connect are valid addresses
 - Validate that bind is "host" or "container" (we should actually turn that into "host" or "guest" and keep "container" as an alias for "guest" internally for backward compatibility)
 - uid, gid security.uid, security.gid should be validated accordingly (valid integer, >= 0)
 - mode should be validated too (valid octal permission)
 - nat and proxy_protocol are boolean

Reply:
That may let you remove a bunch of stuff from the validation you're doing below too.

Reply:
@stgraber I've added those validations now.